Applications/Spire/Source/Ui/TranslatedTableModel.cpp

```diff
@@ -0,0 +1,132 @@
+
+using namespace boost::signals2;
+using namespace Spire;
+
+TranslatedTableModel::ScopeExit::ScopeExit(std::function<void()> f)
+  : m_f(std::move(f)) {}
+
+TranslatedTableModel::ScopeExit::~ScopeExit() {
+  m_f();
+}
+
+TranslatedTableModel::TranslatedTableModel(std::shared_ptr<TableModel> source)
+    : m_source(source),
+      m_transaction_level(0) {
+  m_index_mapping.resize(source->get_row_size());
+  std::iota(m_index_mapping.begin(), m_index_mapping.end(), 0);
+  m_source_operation_connection = m_source->connect_operation_signal(
+    [=] (const Operation& operation) {
+      visit(operation,
+        [=] (const AddOperation& add_operation) {
+          std::transform(m_index_mapping.begin(), m_index_mapping.end(),
+            m_index_mapping.begin(),
+            [&] (auto index) {
+              if(index >= add_operation.m_index) {
+                return index + 1;
+              }
+              return index;
+            });
+          m_index_mapping.push_back(add_operation.m_index);
+          push(AddOperation{get_row_size() - 1});
+        },
+        [=] (const MoveOperation& move_operation) {
+          if(move_operation.m_source < move_operation.m_destination) {
+            std::transform(m_index_mapping.begin(), m_index_mapping.end(),
+              m_index_mapping.begin(),
+              [&] (auto index) {
+                if(index > move_operation.m_source &&
+                    index <= move_operation.m_destination) {
+                  return index - 1;
+                } else if(index == move_operation.m_source) {
+                  return move_operation.m_destination;
+                }
+                return index;
+              });
+          } else {
+            std::transform(m_index_mapping.begin(), m_index_mapping.end(),
+              m_index_mapping.begin(),
+              [&] (auto index) {
+                if(index >= move_operation.m_destination &&
+                    index < move_operation.m_source) {
+                  return index + 1;
+                } else if(index == move_operation.m_source) {
+                  return move_operation.m_destination;
+                }
+                return index;
+              });
+          }
+        },
+        [=] (const RemoveOperation& remove_operation) {
+          auto removed_index = std::distance(m_index_mapping.begin(),
+            m_index_mapping.erase(std::find(m_index_mapping.begin(),
+            m_index_mapping.end(), remove_operation.m_index)));
+          std::transform(m_index_mapping.begin(), m_index_mapping.end(),
+            m_index_mapping.begin(),
+            [&] (auto index) {
+              if(index > remove_operation.m_index) {
+                return index - 1;
+              }
+              return index;
+            });
+          push(RemoveOperation{removed_index});
+        },
+        [=] (const UpdateOperation& update_operation) {
+          push(UpdateOperation{std::distance(m_index_mapping.begin(),
```

Comment:
This is too slow unfortunately, as a general matter the API for this table should not be slower than the API for the source. In other words if the source can implement this in ```O(1)``` then the translation also needs to implement this in ```O(1)``` as well. This implementation, however, is ```O(N)```.

Moves, adds, deletions to the source... those are all ```O(N)``` so it's perfectly fine if this translation is also implements them in ```O(N)```, but updates are universally an ```O(1)``` operation.